### PR TITLE
[SYSTEMDS-3217] PythonAPI working with SYSTEMDS_ROOT set

### PR DESCRIPTION
**Pull request by Baunsgaard**
This commit modify the PythonAPI at startup to work with our release
binaries. While modifying this the building of the python release
artifact have been modified to not include the test and generator code.

### CODE REVIEW

File: src/main/python/pre_setup.py [hunk 1]
```diff
@@ -26,40 +26,56 @@
...
         unzipped_dir_name = file.rsplit('.', 1)[0]
-        shutil.copytree(os.path.join(TMP_DIR, unzipped_dir_name), java_dir_full_path)
-        if os.path.exists(TMP_DIR):
-            shutil.rmtree(TMP_DIR, True)
+        shutil.copytree(os.path.join(TMP_DIR, unzipped_dir_name, 'lib'), LIB_DIR)
+
+# Take hadoop binaries.
+HADOOP_DIR_SRC = os.path.join(SYSTEMDS_ROOT, 'target', 'lib', 'hadoop')
+if os.path.exists(HADOOP_DIR_SRC):
+    shutil.copytree(HADOOP_DIR_SRC, os.path.join(LIB_DIR,"hadoop"))
```
Comment (phaniarnab) [inline src/main/python/pre_setup.py:63]: Please document why do we need the hadoop and winutils dll/libs?

File: src/main/python/pre_setup.py [hunk 2]
```diff
@@ -26,40 +26,56 @@
...
+shutil.copy(os.path.join(SYSTEMDS_ROOT,'conf', 'SystemDS-config-defaults.xml'), os.path.join(this_path, PYTHON_DIR, 'conf', 'SystemDS-config-defaults.xml'))
 
-# delete old build and dist path
+SYSTEMDS_ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.getcwd())))
+shutil.copyfile(os.path.join(SYSTEMDS_ROOT, 'LICENSE'), 'LICENSE')
+shutil.copyfile(os.path.join(SYSTEMDS_ROOT, 'NOTICE'), 'NOTICE')
+
+# Remove old build and dist path
+if os.path.exists(TMP_DIR):
+    shutil.rmtree(TMP_DIR, True)
```
Comment (phaniarnab) [inline src/main/python/pre_setup.py:78]: Changes in the release-related files need testing @j143  and myself, as they can break the release automation.
I would recommend sticking to necessary changes and leaving the improvements for future releases.

**Review by phaniarnab — COMMENTED** [review]
General comment: Please stick to only bug fixes/absolutely necessary changes for the patch release. Code changes in a new RC should only contain the fixes missing in the last RC.

File: src/assembly/bin.xml [hunk 3]
```diff
@@ -91,23 +92,28 @@
...
 				<include>*:log4j*</include>
 				<include>*:netty*</include>
 				<include>*:protobuf-java*</include>
 				<include>*:py4j*</include>
+				<include>*:re2j*</include>
 				<include>*:slf4j-api*</include>
 				<include>*:slf4j-log4j*</include>
 				<include>*:spark-core*</include>
+				<include>*:stax2-api*</include>
+				<include>*:woodstox*</include>
```
[the remark this answers is not in the document]
Comment (Baunsgaard) [reply src/assembly/bin.xml:116]: These are not for the patch fixes of the old release, but they are missing if you make a new release now.

File: src/main/python/pre_setup.py (hunk 1 above)
Comment (Baunsgaard) [reply src/main/python/pre_setup.py:63]: This is for consistancy between release bin version and the python release version.
i will include this in the patch release

File: src/main/python/pre_setup.py (hunk 2 above)
Comment (Baunsgaard) [reply src/main/python/pre_setup.py:78]: yes, that is why i have added the test for the python build back.
i am also testing it for the release artifact.

File: .gitignore [hunk 4]
```diff
@@ -41,19 +41,6 @@ buildNumber.properties
 .jekyll-cache/
 _site/
 
-# Python artifacts
-src/main/python/systemds/systemds-java/
```
Comment (j143) [inline .gitignore:45]: I think it is better only to have one gitignore file for the project.

unless the root .gitignore file is ignoring unintended files in src/main/python!

**Comment by j143** [comment]
Thank you for the cleanup.
Also, we need to create a PR to `branch-2.2.0`!, with specific changes so we can test for spark, java version.

File: .gitignore (hunk 4 above)
Comment (Baunsgaard) [reply .gitignore:45]: i would tend to agree, but in this case the python part is sort of a separate system, that could be in its own repository, and when i develop in the python i teat it as such, therefore i don't have access to the root .gitignore in the IDE when developing the python.

there are examples of doing both designs in Apache.

Hadoop only does one .gitignore
while Spark has one in each language
https://github.com/apache/spark

**Comment by Baunsgaard** [comment]
> Thank you for the cleanup. Also, we need to create a PR to `branch-2.2.0`!, with specific changes so we can test for spark, java version.

yes that is the next task, i will take the appropriate parts from this, and make specific targeted changes for the release branch

**Pull request closed without merging**